### PR TITLE
ocaml 4.12.0

### PR DESCRIPTION
**Pull request by carlocab**
Created with `brew bump-formula-pr`.

### CODE REVIEW

**Comment by carlocab** [comment]
`unison` needs one of https://github.com/bcpierce00/unison/pull/480 or https://github.com/bcpierce00/unison/pull/481. See also https://github.com/bcpierce00/unison/pull/482.

**Comment by rvantonder** [comment]
For `coccinelle`, I believe the compile error seen here:

```
File "stdcompat__arg_s.mli", lines 3-17, characters 0-38:
 3 | type spec = Arg.spec =
 4 |   | Unit of (unit -> unit) 
 5 |   | Bool of (bool -> unit) 
 6 |   | Set of bool ref 
 7 |   | Clear of bool ref 
...
14 |   | Tuple of spec list 
15 |   | Symbol of string list * (string -> unit) 
16 |   | Rest of (string -> unit) 
17 |   | Expand of (string -> string array).
Error: This variant or record definition does not match that of type Arg.spec
       Constructors number 14 have different names, Rest_all and Expand.
```
is being addressed upstream in https://github.com/coccinelle/coccinelle/commit/3dc5d027b448c135959f60cdbea1c77a41c4e4a1
CC @thierry-martinez

**Comment by carlocab** [comment]
https://github.com/coccinelle/coccinelle/commit/3dc5d027b448c135959f60cdbea1c77a41c4e4a1 looks too big to include as a patch. (I'm going to guess it's also not going to apply to the release tarball.)

We may have to wait for a new version of Coccinelle before upgrading to Ocaml 4.12.

**Comment by thierry-martinez** [comment]
I think that installing stdcompat before coccinelle should solve the problem. I tried to propose a formula here: https://github.com/carlocab/homebrew-core/pull/1 but I am not used to homebrew...

**Comment by carlocab** [comment]
Thanks. I don't think we can add `ocaml-stdcompat` to homebrew/core, as CI is going to complain that your repo has too few watchers/stars/forks. I'd be happy to add it to homebrew/core when that changes though.

In the meantime, I think we can install `ocaml-stdcompat` as a bootstrap resource during the build. However, how do we get `opam` and `make` to find it? (It'll be installed in a non-standard location.) Is it enough to add its `bin` to `PATH`, or is it something more involved?

**Comment by carlocab** [comment]
I just tried installing `ocaml-stdcompat` locally, and it seems it's purely a library. I'm actually not super familiar with ocaml, so a hint on how to help `opam`/`make` find it in a non-standard location would be appreciated.

**Comment by gdt** [comment]
unison (not a release yet though) should now build with ocaml 4.12, and does CI with it.   No fixed timetable but we are heading for a release soonish with this - maybe within a month.

**Comment by kit-ty-kate** [comment]
For the `hyperkit` failure feel free to cherry-pick my fix I made for the previous PR bumping the OCaml version (closed without merging since then): https://github.com/chenrui333/homebrew-core/pull/5

**Comment by carlocab** [comment]
Hyperkit is fixed in #72028 but not merged because it's not straightforward to test the build in Homebrew CI.

The only thing blocking this PR is my difficulty here: https://github.com/Homebrew/homebrew-core/pull/71942#issuecomment-790112414

If someone could tell me how I can use `ocaml-stdcompat` in a temporary install whose path I can provide, I can probably sort out the rest of this PR.

**Comment by kit-ty-kate** [comment]
> The only thing blocking this PR is my difficulty here: #71942 (comment)
>
> If someone could tell me how I can use ocaml-stdcompat in a temporary install whose path I can provide, I can probably sort out the rest of this PR.

`export OCAMLPATH=<your-path>` should be enough (I've never had to use it but according to the doc it should be sufficient http://projects.camlcity.org/projects/dl/findlib-1.9.1/doc/ref-html/r865.html)

**Comment by carlocab** [comment]
Thanks, @kit-ty-kate. I've given that a try in e72ce7a3311843db88d669dfee03e6326ee8fd6c.

File: Formula/coccinelle.rb [hunk 1]
```diff
@@ -22,10 +23,25 @@ class Coccinelle < Formula
...
+    sha256 "5e746f68ffe451e7dabe9d961efeef36516b451f35a96e174b8f929a44599cf5"
+  end
+
   def install
+    resource("stdcompat").stage do
+      system "./configure", "--prefix=#{buildpath}/bootstrap"
+      ENV.deparallelize { system "make" }
+      system "make", "install"
+    end
+    ENV.prepend_path "OCAMLPATH", buildpath/"bootstrap"
```
Comment (kit-ty-kate) [inline Formula/coccinelle.rb:43]: if this doesn't work this might need to be buildpath/bootstrap/lib instead, not just buildpath/bootstrap

Comment (carlocab) [reply Formula/coccinelle.rb:43]: Yup, didn't seem to work. I'll wait to see how the Intel nodes fare and then give this a try.

However, I also noticed that `coccinelle` seems to be installing its own ocamlfindlib. Not sure why it's doing that, especially given that it appears to be installing the same version as Homebrew ocaml-findlib.

**Comment by carlocab** [comment]
Unless there are surprises on the Catalina and Big Sur nodes, this should be good to go. Failures on ARM Big Sur and Mojave are expected / pre-existing.

Thanks for all the help, @rvantonder, @thierry-martinez, @gdt, and @kit-ty-kate.

**Pull request closed without merging**